build/xtask/src/dist.rs

```diff
@@ -210,6 +210,101 @@ pub fn list_tasks(app_toml: &Path) -> Result<()> {
 /// Represents allocations and free spaces for a particular image
 type AllocationMap = (Allocations, IndexMap<String, Range<u32>>);
 
+/// Simple data structure to store a set of guaranteed-contiguous ranges
+///
+/// This will panic if you violate that constraint!
+#[derive(Debug, Clone, Default, Hash)]
+pub struct ContiguousRanges(Vec<Range<u32>>);
+impl ContiguousRanges {
+    pub fn new(r: Range<u32>) -> Self {
+        Self(vec![r])
+    }
+    pub fn iter(&self) -> impl Iterator<Item = &Range<u32>> {
+        self.0.iter()
+    }
+    pub fn start(&self) -> u32 {
+        self.0.first().unwrap().start
+    }
+    pub fn end(&self) -> u32 {
+        self.0.last().unwrap().end
+    }
+    pub fn contains(&self, v: &u32) -> bool {
+        (self.start()..self.end()).contains(v)
+    }
+    pub fn push(&mut self, r: Range<u32>) {
+        if let Some(t) = &self.0.last() {
+            assert_eq!(t.end, r.start, "ranges must be contiguous");
+        }
+        self.0.push(r)
+    }
+}
+
+impl<'a> IntoIterator for &'a ContiguousRanges {
+    type Item = &'a Range<u32>;
+    type IntoIter = std::slice::Iter<'a, Range<u32>>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
+}
+
+/// Simple wrapper data structure that enforces that values are ordered
+///
+/// Values must be monotonically increasing or decreasing based on the value of
+/// `self.increasing`.  For example, in the **increasing** mode, each item must
+/// be the same or larger than the previous item.
+struct OrderedVecDeque {
+    data: VecDeque<u32>,
+    increasing: bool,
+}
+impl OrderedVecDeque {
+    /// Build a new `VecDeque` with a decreasing constraint
+    fn decreasing() -> Self {
+        Self {
+            data: VecDeque::new(),
+            increasing: false,
+        }
+    }
+    /// Flip the data and constraint
+    fn reverse(self) -> Self {
```

Comment:
Ok so this is going to sound really pedantic.

Please call this something like `reversed`. Or, since it consumes self, maybe something starting with `into_`?

`reverse` in the standard library is generally an in-place operation (e.g. [on vec](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.reverse)). This caused me to get hung up on how the diff on line 2062 could possibly be correct.

Reply:
Sure, done in 128247d6a06f527a042d11d4a5d0cb4ab3aad528